bee-network/bee-gossip/src/peer.rs

```diff
@@ -0,0 +1,291 @@
+    net::tcp::{OwnedReadHalf, OwnedWriteHalf},
+};
+
+use std::sync::atomic::{AtomicBool, Ordering};
+
+const BUFFER_SIZE: usize = std::mem::size_of::<u32>() + MAX_PACKET_SIZE;
+
+pub(crate) struct PeerInfo {
+    identity: PeerId,
+    alias: String,
+    healthy: AtomicBool,
+}
+
+impl PeerInfo {
+    /// Creates a new connected peer.
+    pub(crate) fn new(identity: PeerId, alias: String) -> Self {
+        Self {
+            identity,
+            alias,
+            healthy: true.into(),
+        }
+    }
+
+    /// The ID of the peer.
+    pub(crate) fn id(&self) -> String {
+        self.identity.id_string()
+    }
+
+    /// The alias of the peer.
+    pub(crate) fn alias(&self) -> &String {
+        &self.alias
+    }
+
+    /// Whether the peer, and therefore the connection is still healthy.
+    pub(crate) fn healthy(&self) -> bool {
+        self.healthy.load(Ordering::Relaxed)
+    }
+}
+
+pub(crate) struct GossipReader {
+    reader: BufReader<OwnedReadHalf>,
+    // FIXME: do we need to preallocate 64Kb for every peer?
+    buffer: Box<[u8; BUFFER_SIZE]>,
+}
+
+impl GossipReader {
+    pub(crate) fn new(reader: BufReader<OwnedReadHalf>) -> Self {
+        Self {
+            reader,
+            buffer: Box::new([0; BUFFER_SIZE]),
+        }
+    }
+
+    pub async fn recv_msgs(&mut self, info: &PeerInfo) -> Result<Vec<(MessageType, Vec<u8>)>, Error> {
+        if info.healthy() {
+            // NOTE:
+            // - every message is prepended by its length: see iotaledger/hive.go/netutil/buffconn/buffconn.go
+            // - Bytes 0..3 encode a u32 representing the message length
+            // - Byte 4     encodes the message type (Message or MessageRequest)
+            // - Bytes 5..n encode the protobuf representation of the actual message
+
+            let num_received = self
+                .reader
+                .read(self.buffer.as_mut())
+                .await
+                .map_err(Error::RecvMessage)?;
+
+            if num_received == 0 {
+                log::warn!("connection reset by peer");
+                info.healthy.store(false, Ordering::Relaxed);
+
+                Err(Error::NotHealthy)
+            } else {
+                log::debug!("received {} bytes", num_received);
+
+                let mut position = 0;
+                let mut messages = Vec::with_capacity(16);
+
+                while position < num_received {
+                    // Determine the length of the next message within this batch
+                    let mut msg_len_buf = [0u8; 4];
+                    msg_len_buf.copy_from_slice(&self.buffer[position..position + 4]);
+                    let msg_len = Buf::get_u32(&mut &msg_len_buf[..]) as usize - 1;
+
+                    // Determine the message type
+                    let msg_type: MessageType =
+                        num::FromPrimitive::from_u8(self.buffer[position + 4]).ok_or(Error::UnknownMessageType)?;
```

Comment:
can we `self.buffer.split` instead to avoid extra bound checks?

Reply:
Done in https://github.com/iotaledger/bee/pull/674/commits/3b5f562211ba7d8de2c217b5c2a7fd86cf8e01dd 

This  is the change that I'm least sure it actually makes anything faster. But the logic here is that any operation with the shape `array[constant]` is optimized away by the compiler and removes the bound check.

Reply:
Understood, but why not just use `get_unchecked`?

Reply:
Right. I think we have no guarantees. Before indexing into it, we could always make sure, that enough bytes (at least 5) are left in the buffer, and thus, prevent any panics and/or UB.